src/v/cloud_storage/inventory/aws_ops.cc

```diff
@@ -108,7 +148,129 @@ ss::future<bool> aws_ops::inventory_configuration_exists(
       {.transfer_details
        = {.bucket = _bucket, .key = _inventory_key, .parent_rtc = parent_rtc},
        .payload = buf});
-    co_return dl_res == download_result::success;
+
+    if (dl_res == download_result::success) {
+        co_return true;
+    }
+
+    if (dl_res == download_result::notfound) {
+        co_return false;
+    }
+
+    co_return error_outcome::failed_to_check_inv_status;
+}
+
+ss::future<op_result<report_metadata>> aws_ops::latest_report_metadata(
+  cloud_storage::cloud_storage_api& remote, retry_chain_node& parent_rtc) {
+    // The prefix is generated according to
+    // https://docs.aws.amazon.com/AmazonS3/latest/userguide/storage-inventory-location.html
+    const auto prefix = cloud_storage_clients::object_key{
+      fmt::format("{}/{}/{}/", _prefix, _bucket, _inventory_config_id)};
+    const auto list_result = co_await remote.list_objects(
+      _bucket, parent_rtc, prefix, '/');
+
+    if (list_result.has_error()) {
+        co_return error_outcome::failed;
+    }
+
+    auto transform_to_checksum = [&prefix](auto datetime) {
+        return checksum_path(prefix().native(), datetime);
+    };
+
+    // The prefix may contain non-datetime folders such as hive data, which we
+    // ignore. The datetime folders are normalized with the prefix, and then the
+    // checksum file path is appended to them. The final result is a series of
+    // paths to checksums which we need to probe.
+    auto filtered = list_result.value().common_prefixes
+                    | std::views::filter(is_datetime)
+                    | std::views::transform(transform_to_checksum);
+
+    std::vector<path_with_datetime> checksum_paths{
+      filtered.begin(), filtered.end()};
+
+    // The datetime strings (YYYY-MM-DDTHH-MMZ) can be sorted by lexical
+    // comparison
+    std::ranges::sort(
+      checksum_paths, [](const auto& first, const auto& second) {
+          return first.datetime > second.datetime;
+      });
+
+    // A checksum appears in the bucket at the designated location only once the
+    // inventory is ready, so we probe the checksum files from latest to
+    // earliest. We stop at the first checksum we find.
+    for (const auto& path_with_datetime : checksum_paths) {
+        if (const auto result = co_await remote.object_exists(
+              _bucket,
+              path_with_datetime.path,
+              parent_rtc,
+              existence_check_type::object);
+            result == download_result::success) {
```

Comment:
It's a bit surprising to proceed in the event this result is an error. Wouldn't that mean we'd proceed with a knowingly stale inventory file?

Reply:
My assumption was that if there is an error, it would be `not_found` because the report is still being prepared.

Additionally the [caller (WIP PR)](https://github.com/redpanda-data/redpanda/pull/17016) is going to compare the returned metadata's datetime with whatever data the caller has last consumed, so it will only every consume newer reports than whatever it last saw, but there is still a problem here. 

If we are encountering a non-404 error even after retries built into remote, then we will probably encounter the same errors for all reports and should probably bail out at the first non 404 error instead of progressively trying older reports.

Reply:
I see, thanks for explaining. Would you mind documenting that assumption in a comment, and/or make sure that if the assumption isn't true that we'll still do something reasonable? For instance, it doesn't necessarily seem impossible for some calls to get unlucky and time out, while subsequent calls succeed.

Reply:
I've made the code path more explicit in the latest push, now if we fail with 404, we try the next latest report. In all other failure cases we abort and the caller can retry later.